File: 2020/20.py
from collections import defaultdict
from functools import reduce
from operator import mul
import sys

def ParseTile(part):
    header, *rows = part.split('\n')
    assert header.startswith('Tile ') and header.endswith(':')
    tile_id = int(header[5:-1])
    return (tile_id, rows)

def GetSides(grid):
    return [grid[0] for grid in Transformations(grid)]

def Rotate(grid):
    return [''.join(grid[c][r] for c in range(len(grid))) for r in range(len(grid[0]) - 1, -1, -1)]

def Flip(grid):
    return [''.join(grid[c][r] for c in range(len(grid))) for r in range(len(grid[0]))]

def Transformations(grid):
    a = grid
    b = Rotate(a)
    c = Rotate(b)
    d = Rotate(c)
    e = Flip(a)
    f = Rotate(e)
    g = Rotate(f)
    h = Rotate(g)
    return [a, b, c, d, e, f, g, h]

def Other(ids, id):
    a, b = ids
    return a if id == b else b

TopSide    = lambda grid: grid[0]
LeftSide   = lambda grid: ''.join(row[0] for row in grid)
BottomSide = lambda grid: grid[-1]
RightSide  = lambda grid: ''.join(row[-1] for row in grid)

# (tile_id -> [row])
tiles = dict(map(ParseTile, sys.stdin.read().strip().split('\n\n')))

# Part 1: find corner tiles, and print product of their ids.
tile_sides = [(tile_id, GetSides(grid)) for (tile_id, grid) in tiles.items()]
tiles_by_side = defaultdict(list)  # side -> [tile_id]
for tile, sides in tile_sides:
    for side in sides:
        tiles_by_side[side].append(tile)

# Sanity check: each side can match at most two tiles (one at the edge)
for side in tiles_by_side.values():
    assert len(side) in (1, 2)

corners = [tile for tile, sides in tile_sides
    if sum(len(tiles_by_side[side]) > 1 for side in sides) == 4]
assert(len(corners) == 4)
print(reduce(mul, corners))


#
# Part 2: reconstruct the image and find the sea monsters!
#

def MakeComposite():
    # Size of the composite grid (i.e. number of tiles along each side).
    # This assumes the result is a square (which it is!)
    S = int(len(tiles)**.5)
    assert S**2 == len(tiles)

    composite_tile_ids = [[None]*S for _ in range(S)]
    composite_subgrids = [[None]*S for _ in range(S)]
    for r in range(S):
        for c in range(S):
            if r == 0 and c == 0:
                # Place topleft corner so that adjacent tiles connect at the
                # right and bottom side. There are two mirrored solutions.
                tile_id = corners[0]
                [grid, _] = [grid for grid in Transformations(tiles[tile_id])
                        if (len(tiles_by_side[RightSide(grid)]) == 2 and 
                            len(tiles_by_side[BottomSide(grid)]) == 2)]
            elif c == 0:
                # Match tile above.
                side = BottomSide(composite_subgrids[r - 1][c])
                tile_id = Other(tiles_by_side[side], composite_tile_ids[r - 1][c])
                grid, = [grid for grid in Transformations(tiles[tile_id]) if TopSide(grid) == side]
            else:
                # Match tile to the left.
                side = RightSide(composite_subgrids[r][c - 1])
                tile_id = Other(tiles_by_side[side], composite_tile_ids[r][c - 1])
                grid, = [grid for grid in Transformations(tiles[tile_id]) if LeftSide(grid) == side]
                if r > 0:
                    assert TopSide(grid) == BottomSide(composite_subgrids[r - 1][c])

            composite_subgrids[r][c] = grid
            composite_tile_ids[r][c] = tile_id

    L = 8  # size of subgrids
    return [''.join(composite_subgrids[r // L][c // L][r % L + 1][c % L + 1]
                for c in range(S * L))
                for r in range(S * L)]

def FindMonsters(grid):
    monster = [
        '..................#.',
        '#....##....##....###',
        '.#..#..#..#..#..#...']

    coords = [(r, c) for r, row in enumerate(monster) for c, ch in enumerate(row) if ch == '#']

    monster_cells = set()
    for r in range(len(grid) - len(monster) + 1):
        for c in range(len(grid[0]) - len(monster[0]) + 1):
            if all(grid[r + i][c + j] == '#' for (i, j) in coords):
                for i, j in coords:
                    monster_cells.add((r + i, c + j))
    return monster_cells

for grid in Transformations(MakeComposite()):
    monster_cells = FindMonsters(grid)
    if monster_cells:
        # Save monsters found for debugging.
        #with open('monsters.txt', 'wt') as f:
        #    for i, row in enumerate(grid):
        #        print(''.join([ch, 'o'][(i, j) in monster_cells] for j, ch in enumerate(row)), file=f)

        # Calculate answer: number of #s in the grid that are not part of a monster.
        answer = sum(ch == '#' and (r, c) not in monster_cells
            for r, row in enumerate(grid) for c, ch in enumerate(row))
        print(answer)

File: 2020/haskell/20.hs
import Data.Function
import Data.List
import Data.List.Split

data Tile = Tile{
    tileId   :: Int,
    topId    :: Int,
    bottomId :: Int,
    leftId   :: Int,
    rightId  :: Int,
    interior :: [String]} deriving Show

-- Rotates a 2D grid clockwise by 90 degrees.
rotate :: [String] -> [String]
rotate xs
    | all null xs = []
    | otherwise   = reverse (map head xs):rotate (map tail xs)

parseTile :: [String] -> (Int, [String])
parseTile (('T':'i':'l':'e':' ':s):rows) | last s == ':' = (read (init s), rows)

-- Given a 2D grid, returns the 8 variations obtained by rotation and mirroring.
transformations :: [String] -> [[String]]
transformations x = [z | y <- [x, reverse x], z <- take 4 (iterate rotate y)]

-- Given a tile id and an initial grid, generates all 8 variations of the tile.
processTile :: (Int, [String]) -> [Tile]
processTile (tileId, grid)
    = [Tile{
            tileId   = tileId,
            topId    = sideId (head grid'),
            bottomId = sideId (last grid'),
            leftId   = sideId (map head grid'),
            rightId  = sideId (map last grid'),
            interior = map (init . tail) $ init $ tail grid'} |
        grid' <- transformations grid]
    where
        -- Interprets a string as a binary number with '#' as 1.
        -- For example, makeId "##.#" == 13.
        sideId :: String -> Int
        sideId = foldl' (\acc ch -> 2*acc + fromEnum (ch == '#')) 0

-- Deletes all variants of `tile` from `tiles` with the same tile id.
deleteTiles :: Tile -> [Tile] -> [Tile]
deleteTiles tile tiles = [t | t <- tiles, tileId t /= tileId tile]

matchTiles :: (Tile -> Tile -> Bool) -> [Tile] -> Tile-> ([Tile], [Tile])
matchTiles match remainingTiles tile = (remainingTiles', tile:tiles')
    where
        (remainingTiles', tiles') = case [t | t <- remainingTiles, match tile t] of
            []         -> (remainingTiles, [])
            [nextTile] -> matchTiles match (deleteTiles nextTile remainingTiles) nextTile
            _          -> error "multiple matches possible"

createColumn :: [Tile] -> Tile -> ([Tile], [Tile])
createColumn = matchTiles (\a b -> bottomId a == topId b)

createRow :: [Tile] -> Tile -> ([Tile], [Tile])
createRow = matchTiles (\a b -> rightId a == leftId b)

-- `stitchRows remainingTiles firstColumn` expands the first column into a full grid of tiles,
-- using `createRow` to expand the leftmost tile of each row into a full row of tiles.
--
-- We could double-check that tiles in columns beyond the first also match along up/down edge,
-- but for now we'll assume that's the case.
stitchRows :: [Tile] -> [Tile] -> ([Tile], [[Tile]])
stitchRows remainingTiles [] = (remainingTiles, [])
stitchRows remainingTiles (tile:tiles) = (remainingTiles'', row:rows)
    where
        (remainingTiles', row) = createRow remainingTiles tile
        (remainingTiles'', rows) = stitchRows remainingTiles' tiles

identifyMonsters :: [String] -> [String]
identifyMonsters []   = []
identifyMonsters grid = identifyHorizontal (head grid : identifyMonsters (tail grid))
    where
        identifyHorizontal :: [String] -> [String]
        identifyHorizontal grid
            | all null grid = grid
            | otherwise     = identifyHere (
                zipWith (:) (map head grid) (identifyHorizontal (map tail grid)))

        identifyHere :: [String] -> [String]
        identifyHere grid
            | detectMonster grid = imposeMonster grid
            | otherwise          = grid

        detectMonster :: [String] -> Bool
        detectMonster = matchAll monster
            where
                matchAll [] _ = True
                matchAll _ [] = False
                matchAll (a:monster) (b:grid) = matchRow a b && matchAll monster grid

                matchRow []          _       = True
                matchRow _           []      = False
                matchRow (a:monster) (b:row) =
                    (a /= '#' || b == '#'|| b == 'O') && matchRow monster row

        imposeMonster :: [String] -> [String]
        imposeMonster = imposeAll monster
            where
                imposeAll []          grid     = grid
                imposeAll (a:monster) (b:grid) = imposeRow a b:imposeAll monster grid

                imposeRow []          row     = row
                imposeRow (a:monster) (b:row) =
                    (if a == '#' then 'O' else b):imposeRow monster row

        monster :: [String]
        monster = [
            "                  # ",
            "#    ##    ##    ###",
            " #  #  #  #  #  #   "]

main = do
    input <- getContents
    let tiles = concatMap (processTile . parseTile) $ filter (/= []) $ splitOn [[]] $ lines input

    -- Find corner tiles (note: each corner appears twice in the list due to mirroring!)
    let corners = [t | t <- tiles,
            null [t' | t' <- tiles,
                    tileId t /= tileId t', topId t == bottomId t' || leftId t == rightId t']]

    -- Part 1: print product of corner tile ids
    print $ product $ nub $ map tileId corners

    -- Part 2: reconstruct grid and find monsters.
    let corner = head corners :: Tile
    let ([], tileGrid) = uncurry stitchRows $ createColumn (deleteTiles corner tiles) corner
    let compositeGrid = concatMap (map concat . transpose . map interior) tileGrid :: [String]
    -- There should be exactly one orientation that has some monsters in it.
    let [orientedGrid] = [grid |
            grid <- map identifyMonsters $ transformations compositeGrid, 'O' `elem` concat grid]
    print $ length $ filter ('#' ==) $ concat orientedGrid
